### PR TITLE
Issue/8011 dryrun facts

### LINKED ISSUE #8011: resource scheduler: implement dry-run and refresh-fact
Make sure the associated tasks have all relevant fields + implement on the agent side. The server is still responsible for the timer and sending the refresh-fact trigger to the scheduler.

#7561 may already contain parts of this, though it was created before the resource scheduler design

### PR DESCRIPTION
**Pull request by wouterdb**
# Description

Added dryrun and get facts 

closes #8011 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

### CODE REVIEW

File: tests/agent_server/deploy/e2e/test_dryrun_new.py [hunk 1]
```diff
@@ -0,0 +1,563 @@
...
+        "from_value_compare": json.dumps({"first_level": {"nested": [5, 3, 2, 1, 1]}}, indent=4, sort_keys=True),
+        "to_value": resources[3]["value"],
+        "to_value_compare": json.dumps(resources[3]["value"], indent=4, sort_keys=True),
+    }
+    assert changes[4]["status"] == "unmodified"
+    assert changes[5]["status"] == "undefined"
+    assert changes[6]["status"] == "skipped_for_undefined"
+    assert changes[7]["status"] == "skipped_for_undefined"
+
+    pytest.skip("Agent down will only work when we have pause ability!!!")
```
Comment (hlloreda) [inline tests/agent_server/deploy/e2e/test_dryrun_new.py:538]: Do we have a ticket to remove this line when we have the pause ability? Or is it only a `FIXME`?

File: tests/agent_server/deploy/e2e/test_facts_new.py [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+    assert result.code == 200
+
+    await _wait_until_deployment_finishes(client, environment, version)
+
+    result = await client.get_version(environment, version)
+    assert result.result["model"]["done"] == len(resources)
+
+    # The resource facts should be purged
+    result = await client.get_param(environment, "length", resource_id_wov)
+    assert result.code == 503
```
Comment (hlloreda) [inline tests/agent_server/deploy/e2e/test_facts_new.py:128]: I'm not sure to understand how this proves that the fact has been purged, I only see this in the result:
```
'Agents queried for resource parameter.'
```

File: tests/agent_server/deploy/e2e/test_facts_new.py [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+
+    result = await client.release_version(environment, version, True, const.AgentTriggerMethod.push_full_deploy)
+    assert result.code == 200
+
+    await _wait_until_deployment_finishes(client, environment, version)
+
+    await get_fact("test::Fact[agent1,key=key1]", 503, lower_limit=0)  # undeployable
+    await get_fact("test::Fact[agent1,key=key2]")  # normal
+    await get_fact("test::Fact[agent1,key=key3]")  # not present -> present
+    pytest.skip("No unknowns yet!")
```
Comment (hlloreda) [inline tests/agent_server/deploy/e2e/test_facts_new.py:284]: Same comment for the previous skip

**Review by hlloreda — CHANGES_REQUESTED** [review]

File: src/inmanta/deploy/tasks.py [hunk 4]
```diff
@@ -170,10 +188,44 @@ async def report_deploy_failure(excn: Exception) -> None:
...
         return False
 
+    async def execute(self, scheduler: "scheduler.ResourceScheduler", agent: str) -> None:
+        try:
+            my_executor: executor.Executor = await self.get_executor(
+                scheduler, agent, self.resource_full.id.entity_type, self.version
+            )
+            await my_executor.dry_run([self.resource_full], self.dryrun_id)
+        except Exception:
+            logger_for_agent(agent).error(
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/deploy/tasks.py:204]: There is no such thing, `.exception` is the same as `.error(exc_info=True)`

File: tests/agent_server/deploy/e2e/test_dryrun_new.py (hunk 1 above)
Comment (wouterdb) [reply tests/agent_server/deploy/e2e/test_dryrun_new.py:538]: We should indeed make sure we have no more skips at the end, I'll ad it the ticket.

File: tests/agent_server/deploy/e2e/test_facts_new.py (hunk 2 above)
Comment (wouterdb) [reply tests/agent_server/deploy/e2e/test_facts_new.py:128]: I copied pasted this testcase to ensure the behavior is the same.

This at least shows it is dropped from the DB (it was 200, it hasn't timed out, now it is gone,...)

File: tests/agent_server/deploy/e2e/test_facts_new.py (hunk 3 above)
Comment (wouterdb) [reply tests/agent_server/deploy/e2e/test_facts_new.py:284]: added to https://github.com/inmanta/inmanta-core/issues/8013

File: src/inmanta/deploy/tasks.py [hunk 5]
```diff
@@ -170,10 +188,44 @@ async def report_deploy_failure(excn: Exception) -> None:
 @dataclass(frozen=True, kw_only=True)
 class DryRun(Task):
     version: int
+    resource_full: state.ResourceDetails
```
Comment (sanderr) [inline src/inmanta/deploy/tasks.py:191]: Perhaps simply `details` instead of `resource_full`?

File: src/inmanta/deploy/scheduler.py [hunk 6]
```diff
@@ -119,22 +121,42 @@ async def repair(self) -> None:
...
+                        resource=rid,
+                        version=version,
+                        resource_full=resource,
+                        dryrun_id=dry_run_id,
+                    ),
+                    priority=10,
+                )
+            )
 
     async def get_facts(self, resource: dict[str, Any]) -> None:
```
Comment (sanderr) [inline src/inmanta/deploy/scheduler.py:138]:
```suggestion
    async def get_facts(self, resource: dict[str, object]) -> None:
```

Not from this PR, but I only just noticed it.

File: src/inmanta/deploy/scheduler.py [hunk 7]
```diff
@@ -119,22 +121,42 @@ async def repair(self) -> None:
...
-        pass
+        resources = await self.build_resource_mappings_from_db(version)
+        for rid, resource in resources.items():
+            self._work.agent_queues.queue_put_nowait(
+                PrioritizedTask(
+                    task=DryRun(
+                        resource=rid,
+                        version=version,
+                        resource_full=resource,
+                        dryrun_id=dry_run_id,
```
Comment (sanderr) [inline src/inmanta/deploy/scheduler.py:132]: Small inconsistency in naming here.

**Review by sanderr — CHANGES_REQUESTED** [review]

**Review by hlloreda — APPROVED** [review]

**Review by sanderr — APPROVED** [review]

**Pull request closed without merging**